python/tank/pipelineconfig_factory.py

```diff
@@ -188,104 +162,229 @@ def _from_path(path, force_reread_shotgun_cache):
+
         return PipelineConfiguration(pc_registered_path)
 
-    # now get storage data, use cache unless force flag is set
+    # now get storage and project data from shotgun.
+    # this will use a cache unless the force flag is set
     sg_data = _get_pipeline_configs(force_reread_shotgun_cache)
 
     # now given ALL pipeline configs for ALL projects and their associated projects
     # and project root paths (in sg_data), figure out which pipeline configurations
-    # are matching the given path.
-
+    # are matching the given path. This is done by walking upwards in the path
+    # until a project root is found, and then figuring out which pipeline configurations
+    # belong to that project root.
     associated_sg_pipeline_configs = _get_pipeline_configs_for_path(path, sg_data)
 
+    log.debug(
+        "Associated pipeline configurations are: %s" % pprint.pformat(associated_sg_pipeline_configs)
+    )
+
     if len(associated_sg_pipeline_configs) == 0:
-        # no matches! The path is unknown or invalid.
-        raise TankError("The path '%s' does not seem to belong to any known Toolkit project!" % path)
+        # no matches! The path is invalid or does not belong to any project on the current sg site.
+        raise TankInitError(
+            "The path '%s' does not belong to any known Toolkit project!" % path
+        )
+
+    # perform various validations to make sure the version of the sgtk codebase running
+    # is associated with the given configuration correctly, and if successful,
+    # create a pipeline configuration
+    return _validate_and_create_pipeline_configuration(
+        associated_sg_pipeline_configs,
+        source=path
+    )
+
+
+def _validate_and_create_pipeline_configuration(associated_pipeline_configs, source):
+    """
+    Given a set of pipeline configuration, validate that the currently running code
+    is compliant and construct and return a suitable pipeline configuration instance.
+
+    This method takes into account complex new and old logic, including classic
+    setups, shared cores and bootstrap workflows.
+
+    The associated_pipeline_configs parameter contains a list of potential pipeline
+    configuration shotgun dictionaries which should be considered for resolution.
+    Each dictionary contains the following entries:
+
+        - id (pipeline configuration id)
+        - type (e.g. ``PipelineConfiguration``)
+        - code
+        - windows_path
+        - linux_path
+        - mac_path
+        - project (associated project entity link)
+        - project.Project.tank_name (associated project tank name)
 
-    # extract current os path data from the pipeline configuration shotgun data
-    (all_pc_data, primary_pc_data) = _get_pipeline_configuration_data(associated_sg_pipeline_configs)
+    Thism method will either return a pipeline configuration instance based on
+    one of the ``associated_pipeline_configs`` entries or raise a TankInitError
+    exception.
```

Comment:
maybe just add the reason why the exception would be raised here? i.e. validation fails.